### PR TITLE
Update knn query spec

### PR DESCRIPTION
**Pull request by alex-keeler**
### Description
Updates the knn query spec to match how it is currently implemented in OpenSearch.

### Issues Resolved
Closes https://github.com/opensearch-project/opensearch-api-specification/issues/535

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by alex-keeler** [comment]
The change was a bit more involved than I had originally anticipated.
Based off what I can tell, it appears the existing spec was ripped from the equivalent Elasticsearch spec. The OpenSearch implementation differs pretty significantly from this spec -- notably that the `fields` parameter has been replaced with a nested object, which uses the field name as the key. This nested object is what contains the rest of the parameters. Such behavior provided some challenges in designing the OpenAPI spec, so if there's a better alternative to my solution I'd be happy to update it. I also split the API between the KnnQuery and a new KnnField component to try and simplify the nested object business -- that can also be consolidated if deemed necessary.

File: spec/schemas/_common.yaml [hunk 1]
```diff
@@ -591,23 +591,19 @@ components:
...
           $ref: '#/components/schemas/QueryVector'
-        query_vector_builder:
-          $ref: '#/components/schemas/QueryVectorBuilder'
         k:
           description: The final number of nearest neighbors to return as top hits
           type: number
-        num_candidates:
-          description: The number of nearest neighbor candidates to consider per shard
+        min_score:
+          description: The minimum similarity score for a neighbor to be considered a hit
```
Comment (dblock) [inline spec/schemas/_common.yaml:603]: Nit: add periods at the end of these descriptions.

**Review by dblock — CHANGES_REQUESTED** [review]
👏 great work

- Update CHANGELOG.
- Annotate the APIs with `x-version-added` and such, 1.3.x tests fail with "[knn] query does not support [min_score])" for example.

@nhtruong @Xtansia might have some suggestions around the spec itself.

**Review by dblock — DISMISSED** [review]

**Comment by dblock** [comment]
For lints, probably just easier to sort values.

```
/home/runner/work/opensearch-api-specification/opensearch-api-specification/tests/default/_core/search/knn.yaml
Error:    36:31  error  Expected sequence values to be in ascending order. '2.3' should be before '3.5'  yml/sort-sequence-values
```

Tests also need to be limited to `>= 2.14` with `version:`, it's not smart enough to know that you're trying to test APIs that were added later to skip them.

**Comment by dblock** [comment]
Looks like the filter query doesn't work in these old versions either. Not sure whether that's expected.

```
[INFO] => POST /movies/_search ({}) [application/json] {
  "query": {
    "knn": {
      "embedding": {
        "vector": [
          1.4,
          2.3,
          3.5,
          4.1,
          9.2
        ],
        "k": 1,
        "filter": {
          "term": {
            "year": 2012
          }
        }
      }
    }
  }
}
[INFO] <= 400 (application/json) | {
  "root_cause": [
    {
      "type": "parsing_exception",
      "reason": "[knn] unknown token [START_OBJECT] after [filter]",
      "line": 1,
      "col": 77
    }
  ],
  "type": "parsing_exception",
  "reason": "[knn] unknown token [START_OBJECT] after [filter]",
  "line": 1,
  "col": 77
}
```

You can reproduce these failures locally pretty easily btw.

```
OPENSEARCH_VERSION=1.3.7 docker compose up
```

```
npm run test:spec--insecure -- --tests tests/default/_core/search/knn.yaml --verbose
```

**Comment by alex-keeler** [comment]
Thanks for the heads up, I've only been testing against 2.16.0 locally. I'll go through and make sure all is well with earlier releases. I'm sure the failures are as annoying for you as they are for me.
I'll have to do some digging to see when the filter parameter was added. I didn't realize it came in a later release (I'm assuming, anyway).

**Comment by dblock** [comment]
@navneet1v from k-nn might be interested in this and can help make sure things are properly version-annotated

**Comment by navneet1v** [comment]
> Looks like the filter query doesn't work in these old versions either. Not sure whether that's expected.
> 
> ```
> [INFO] => POST /movies/_search ({}) [application/json] {
>   "query": {
>     "knn": {
>       "embedding": {
>         "vector": [
>           1.4,
>           2.3,
>           3.5,
>           4.1,
>           9.2
>         ],
>         "k": 1,
>         "filter": {
>           "term": {
>             "year": 2012
>           }
>         }
>       }
>     }
>   }
> }
> [INFO] <= 400 (application/json) | {
>   "root_cause": [
>     {
>       "type": "parsing_exception",
>       "reason": "[knn] unknown token [START_OBJECT] after [filter]",
>       "line": 1,
>       "col": 77
>     }
>   ],
>   "type": "parsing_exception",
>   "reason": "[knn] unknown token [START_OBJECT] after [filter]",
>   "line": 1,
>   "col": 77
> }
> ```
> 
> You can reproduce these failures locally pretty easily btw.
> 
> ```
> OPENSEARCH_VERSION=1.3.7 docker compose up
> ```
> 
> ```
> npm run test:spec--insecure -- --tests tests/default/_core/search/knn.yaml --verbose
> ```

The filter query specifically provided here doesn't work for old version because filters works for 2.4 and above.

**Comment by alex-keeler** [comment]
> > Looks like the filter query doesn't work in these old versions either. Not sure whether that's expected.
> > ```
> > [INFO] => POST /movies/_search ({}) [application/json] {
> >   "query": {
> >     "knn": {
> >       "embedding": {
> >         "vector": [
> >           1.4,
> >           2.3,
> >           3.5,
> >           4.1,
> >           9.2
> >         ],
> >         "k": 1,
> >         "filter": {
> >           "term": {
> >             "year": 2012
> >           }
> >         }
> >       }
> >     }
> >   }
> > }
> > [INFO] <= 400 (application/json) | {
> >   "root_cause": [
> >     {
> >       "type": "parsing_exception",
> >       "reason": "[knn] unknown token [START_OBJECT] after [filter]",
> >       "line": 1,
> >       "col": 77
> >     }
> >   ],
> >   "type": "parsing_exception",
> >   "reason": "[knn] unknown token [START_OBJECT] after [filter]",
> >   "line": 1,
> >   "col": 77
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You can reproduce these failures locally pretty easily btw.
> > ```
> > OPENSEARCH_VERSION=1.3.7 docker compose up
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > npm run test:spec--insecure -- --tests tests/default/_core/search/knn.yaml --verbose
> > ```
> 
> The filter query specifically provided here doesn't work for old version because filters works for 2.4 and above.

Thank you for the info. I ended up restraining the filter test to >= 2.9, as it appears the faiss engine didn't get filter support until then. I tried switching the embedding to lucene, but found out that lucene knn support wasn't added until version 2.2, so I figured it was better to keep things as is. For extra safety, I also constrained the entire test to >= 1.2, which is when faiss k-NN support was added.

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
Thanks so much for your help @alex-keeler. Please don't hesitate to pickup other missing/potentially incorrect parts of the search DSL, that would be much appreciated!

**Pull request merged**